extensions/interactions/GraphInput/GraphInput.js

```diff
@@ -583,6 +583,86 @@ oppia.directive('graphViz', function() {
 
 oppia.factory('graphUtilsService', [function() {
   return {
+    GRAPH_ADJACENCY_MODE: {
+      DIRECTED: 'directed',
+      INVERTED: 'inverted',
+      UNDIRECTED: 'undirected'
+    },
+    DFS_STATUS: {
+      VISITED: 'visited',
+      UNVISITED: 'unvisited',
+      STILL_VISITING: 'still visiting'
+    },
+    /**
+     * @param {object} graph A graph object.
+     * @param {string} mode A string indicating the mode.
+     * @return {array} An adjacency list. Depending on the mode, the list has
+     *   all edges (directed),
+     *   all edges inverted (inverted),
+     *   or all edges in both directions, as though the graph were undirected
+     *   (undirected)
+     */
+    constructAdjacencyLists: function(graph, mode) {
+      var adjacencyLists = [];
+      for (var i = 0; i < graph.vertices.length; i++) {
+        adjacencyLists.push([]);
+      }
+
+      // If a graph is undirected, all modes work the same way anyway
+      if (!graph.isDirected) {
+        mode = this.GRAPH_ADJACENCY_MODE.UNDIRECTED;
+      }
+      for (var i = 0; i < graph.edges.length; i++) {
+        var edge = graph.edges[i];
+        if (mode == this.GRAPH_ADJACENCY_MODE.DIRECTED ||
+            mode == this.GRAPH_ADJACENCY_MODE.UNDIRECTED) {
+          adjacencyLists[edge.src].push(edge.dst);
+        }
+        if (mode == this.GRAPH_ADJACENCY_MODE.INVERTED ||
+            mode == this.GRAPH_ADJACENCY_MODE.UNDIRECTED) {
+          adjacencyLists[edge.dst].push(edge.src);
+        }
+      }
+      return adjacencyLists;
+    },
+    /**
+     * @param {integer} startVertex The index of the starting vertex.
+     * @param {array} adjacencyLists An array of arrays.
+     * @param {array} isVisited An array with length equal to the number of
+     *     vertices. All the values should be false initially.
+     * This function modifies the isVisited array and changes the values at
+     * the indices of the vertices reachable form the starting vertex to true.
+     */
+    markVisited: function(startVertex, adjacencyLists, isVisited) {
```

Comment:
What do you have in mind for a brief description that's different from the docstring?